content/authentication/reading/hashing-passwords/_index.md

```diff
@@ -0,0 +1,148 @@
+The diagram below shows the general flow of how hashing works to secure the site.
+
+![Diagram showing how encoded passwords are compared to a submitted password to authenticate a user.](pictures/hashingdiagram.png?classes=border)
+
+It can be difficult to write a good hash function, so thankfully you will never have to. Over the years, some very smart people have created some very good hash functions. However, it is important to understand how the properties of a hash function allow for secure password storage. For our purposes, it is only important that we consider a hash function to have these properties:
+
+### Hash Properties
+
+1. **Deterministic**: If we encode a message with the function at different points in time then we always get the same result.
+1. **Hard to reverse**: It is infeasible to calculate the input value that yields a given hash.
+1. **Hash values are ALMOST unique**: If `a` and `b` are two different messages, then it is *extremely* unlikely that they have the same hash value. By "extremely unlikely" we mean that this is something like a one-in-a-trillion likelihood (for example, there are only about 8 billion people on earth). 
+1. **Similar messages have VERY different hash values**: In other words, if we change a message only slightly, the resulting hash value is *very* different. A function that returned `AlXL3M_ws` for the message `"LaunchCode"` and `AlXL3M_wt` for the message `"LaunchCodf"` would not be a suitable hash function. 
+
+### Password Verification With Hashes
+
+Our application will select a particular hash function. Let's call it `h`. Then, for a message `x`, the hash value will be the result of calling `h` with the argument `x`. Invoked, this looks like `h(x)`.  Rather than store passwords in a database, we will store their hash values.
+
+Consider a fictional user that wants to sign up for our site, Jamie. Jamie likes Taylor Swift, so their desired username is `tswizzle_fan` and their desired password is `lover1989` (not a great password choice, by the way, but we've seen worse).
+
+When Jamie registers for an account on our site, we will call our hash function with their password:
+
+```csharp {linenos=table}
+   string hash = h("lover1989"); // returns "bd62d7e13ef465fa705f30de198ba0cb"
+
+```
+
+Then we store Jamie's username along with her hashed password in our `user` table.
+
+| username | passwordHash |
+|-----|-----|
+| tswizzle_fan | bd62d7e13ef465fa705f30de198ba0cb |
+
+This is very secure. Even if somebody breaks into the database and finds Jamie's info, they will not be able to log in. Since we used a hash function, the hacker will have a very hard time turning the hash into a password:[see property 2](http://localhost:1313/authentication/reading/hashing-passwords/#hash-properties)
```

Comment:
```suggestion
This is very secure. Even if somebody breaks into the database and finds Jamie's info, they will not be able to log in. Since we used a hash function, the hacker will have a very hard time turning the hash into a password:[see property 2](#hash-properties)
```